### ISSUE TITLE
Responsive layout

### ISSUE THREAD

**Issue by metasoarous** [matsengrp/olmsted#28]
Should follow nextstrain in making the layout responsive in order to take as much advantage of screen real estate as possible. The clonal families viz and table in particular should expand to fit the width/height of the screen, and the tree-alignment viz should probably split the screen width, possibly with a movable slider.

**Comment by eharkins** [comment]
This is taken from the nextstrain function that helps resize components @ https://github.com/nextstrain/auspice/blob/master/src/util/computeResponsive.js
``` 
This function figures out the dimensions of a panel. There are several challenges to this.
  It's harder mental model wise than code wise.
  In a normal responsive layout, you would allow flexbox (which nextstrain uses) to determine
  the width of an element. In that case, one describes a layout with a higher level API and
  lets the browser compute the widths and heights.
  That doesn't work because most of the panels have svgs that have scales, and scales need absolute
  width and height values.
```
I'm wondering if vega is immune to this scale problem? I think its coordinates are only important in the context of its container, though I could be wrong. More investigation required, but we have props related to available width and height from nextstrain in our top level Contents component, and I am hoping they just need to be passed down and used to determine flexbox proportions. There are of course questions associated with things like the tree that can resize themselves.

**Comment by metasoarous** [comment]
I think it just means that an SVG or Canvas isn't naturally able to fill available width the way a div or other container is, because there are things that need to be computed from exact width/height specifications. In general though, I think you're right; vega does sort of paper over these concerns for us a bit, as whatever width/height we pass in will flush through the system and produce a new viz. I'm not sure how different that looks for the d3; probably not that much different really. Bottom line is its not like saying "width: 70%", like it would be for other elements. But still, not hard, just have to do our own computations for accounting for padding etc using the provided function.

You're right that the real question is what we do with the resizable tree. My vote is:
* force the tree / alignment to split the screen width 30/70% (or thereabouts)
* have a movable boundary between the two that updates the widths as you move the boarder around
* if you want to "zoom" on the tree, maybe make that possible width mouse wheel or something

**Comment by eharkins** [comment]
Using the zoom feature seems like it would eliminate the need for height and branch scales altogether as it seems to serve the same purpose (of navigating a tree that doesn't fit into its container) in a smoother interaction. The only question that seems difficult to answer until we try, is will it complicate things in terms of lining the tree up on the y axis with the alignment. But hopefully @metasoarous' idea of restricting the zoom feature to the x axis will work, and even better if we are able to make it work with the y axis and the alignment!

**Comment by eharkins** [comment]
Also important to consider any constraints introduced by #28. That is, we want to be able to highlight a sequence in one viz when hovering on it in its adjacent partner, and this is worth considering as we think about how we are navigating the tree and which sequences are visible at any point (if we have done some zooming, for example).

**Comment by metasoarous** [comment]
Yes, this keeping the y-axis aligned is why zoom should be restricted to the x-axis. Not sure there's anything left to think through with #28 as long as we do it this way. As long as something is visible we should be able to hover on it, and if you want to hover on something outside zoom, you just have to rezoom.

You're also right that this does potentially remove the need for branch scales, if a combination of zooming and resizing does the job. Again, it would still be nice to be able to adjust the boundary between the tree and the alignment by dragging, but I guess this would have to happen in the vega, which might be a bit tricky. And it would be doubly nice to also be able to click drag to resize in the vertical direction as well.

I took a look at the zoom example in greater detail, and it's definitely involved if you want to be able to drag the visible area around, which I think we'd probably want. It may be simpler restricted to the x-axis though.

So lets see what happens here; if all of the details here end up being a bit tricky, I think it's fine to get it this part of the viz passable (one way or the other), and open up a second issue to smooth out the details.

**Comment by eharkins** [comment]
> Also important to consider any constraints introduced by #24. 

As you probably got from the context, I meant #24 (not #28 which is this issue), and I've updated the comment to reflect that.

>As long as something is visible we should be able to hover on it, and if you want to hover on something outside zoom, you just have to rezoom.

If the hovering goes in both directions, that is, if we can hover on an alignment and want to highlight the corresponding leaf in the tree, there is a possibility that it isn't visible because of how the tree is currently zoomed. We could unzoom the tree if that is the case, do nothing, or maybe we only care about highlighting in the tree to show the corresponding alignment. There could be other options too, but it's worth considering.

This all sounds good, I'll get started on this soon and see where the challenges lie.

**Comment by metasoarous** [comment]
It would be cool to do a snap-to thing for the tree if you hover over a sequence in the alignment but the corresponding tree node isn't showing. But I certainly don't think this is necessary to start. If someone has started zooming around, they've opted themselves into the situation and can get out of it easily enough. We can certainly open up a separate issue for this for when we have more time though.

**Comment by metasoarous** [comment]
For now, let's do the % width split between tree and alignment based on a slider, as you suggested. That will be easy enough and solve the problem well enough for now, while setting things up so we can introduce a draggable barrier down the road. I'll open a separate issue for getting the draggable barrier.

**Comment by eharkins** [comment]
@metasoarous What would be the method of expanding the tree and looking at a particular clade with out zoom on the x-axis? Right now, the tree viz resizes itself and pushes over the alignment when you expand the branch-length scale with the slider. But if this stays the same and we want to contain the tree viz within a resizable box controlled by an additional slider as we've been discussing, then does the user need to slide the branch scale out and then slide the tree viz slider to be able to see a zoomed in tree? Let me know if this makes sense or not, just wanted to clarify the desired behavior for end-user if we are keeping it simple (without zoom).

**Comment by metasoarous** [comment]
What I've been trying to explain here is that everything should be computed based on the width of the visualization. So as we change the dimensions of the screen, different width values get passed down as props to our vega component, and the vega uses those width values as signals to figure out how to render the tree in the allotted space. That is to say, the branch length scale should be computed from the width, not vice versa as it is now.

> What would be the method of expanding the tree and looking at a particular clade with out zoom on the x-axis?

You just move the slider over to give the tree 80% of the width instead of 30% of the width (default). Or maximize your screen.

> Right now, the tree viz resizes itself and pushes over the alignment when you expand the branch-length scale with the slider. But if this stays the same and we want to contain the tree viz within a resizable box controlled by an additional slider as we've been discussing, then does the user need to slide the branch scale out and then slide the tree viz slider to be able to see a zoomed in tree?

No, of course not. This is what I mean by width holding primacy over tree scale. Tree scale as it stands now should probably be derived from width and a new signal `zoom`, which will be 1 if the tree should take up the whole width, 0.5 if it should take up 1/2 the width, and 2 to zoom in 2x. Tree scale shouldn't be a control. Zoom should be a control. Tree scale should just be what we compute to make the selected width and zoom make sense.

Hypothesis: Part of the problem is that vega thinks the viz is one size, but as the paths start getting drawn, they fall outside the expected region, and so the svg resizes itself to try and roll with the new marks, but Vega doesn't see these new _actual_ widths and heights, and so thinks everything looks very different than it actually does.

Evidence: This hypothesis offers an explanation for the extra space below the viz (#43), the weird way that the alignment doesn't always rerender of fall back into place when we shrink the branch scale back down, and the bizarre axis label crunch (#60).

If I'm correct here, basing everything on the width, and focusing on changing the relative width of each part of the viz as a function of that, should fix everything. And even if this hypothesis is wrong, this really seems like the right way to organize things, and hopefully at least makes the mess more comprehensible.

I'm happy to let you take this part on if you're keen @eharkins, but let me know if you want me to jump on this part of the issue, as I think I have an idea about the direction to go. Again though, if for starters we can just get the main clonal families viz and table in place, that will get us a good ways forward on this issue, and set up precedent for the screen dimension data flow.

**Comment by eharkins** [comment]
This makes sense.  @metasoarous, I will start with just getting the clonal familiez viz width sorted out and we can go from there, feel free to get started on the tree part if you want, otherwise I'll move on to that when I get the initial viz working.

**Comment by eharkins** [comment]
@metasoarous Take a look at the above commit when you get a chance in `28-resizable-vega` branch. This issue has of course spawned many others  (#58 #59 #63 #64 #65 etc) so when we merge `28-resizable-vega` should we close this issue?

**Comment by metasoarous** [comment]
Left a review with a few comments; but mostly looks good!

One thing we also want to get if possible is a max height on the main clonal families viz; This is suboptimal for mobile: 
![image](https://user-images.githubusercontent.com/88556/46331999-4321f080-c5cf-11e8-814b-867df09350e8.png)

**Comment by eharkins** [comment]
@metasoarous Just setting the height to be 0.75 times width for now

**Comment by metasoarous** [comment]
@eharkins As we talked about elsewhere, it sounds like the issue around dynamic height resizing has been resolved, and I think we should do that instead of fixed 0.75. The reason is that for the more customized vizs (e.g. #34), it will be really nice to let it take up as much screen real estate as possible.

**Comment by metasoarous** [comment]
Yes, I think that's fine for now.
